File: pyclient/confluo/rpc/type_conversions.py
from data_types import DataType
from schema import SchemaBuilder, Schema
from ttypes import rpc_column


def convert_to_rpc_schema(schema):
    """ Converts a schema to an rpc schema for the client.

    Args:
        schema: The schema to convert.
    Returns:
        The rpc schema.
    """
    rpc_columns = []
    for column in schema.columns_:
        rpc_col = rpc_column(column.data_type_.type_id_, column.data_type_.size_, column.name_)
        rpc_columns.append(rpc_col)
    return rpc_columns


def convert_to_schema(rpc_schema):
    """ Converts an rpc schema to a schema for confluo.

    Args:
        rpc_schema: The rpc_schema to convert.
    Returns:
        The schema for confluo.
    """
    builder = SchemaBuilder()
    for column in rpc_schema:
        builder.add_column(DataType(column.type_id, column.type_size), column.name)
    return Schema(builder.build())

File: javaclient/src/main/java/confluo/rpc/TypeConversions.java
package confluo.rpc;

import java.util.ArrayList;
import java.util.List;

/**
 * Handler of conversions between RPC and non RPC types
 */
class TypeConversions {

  /**
   * Converts a schema to an rpc schema for the client
   *
   * @param schema The schema to convert
   * @return The rpc schema
   */
  static List<rpc_column> convertToRPCSchema(Schema schema) {
    List<rpc_column> rpcColumns = new ArrayList<>();
    for (Column column : schema.getColumns()) {
      rpc_column rpcCol = new rpc_column(column.getDataType().typeId.getValue(), column.getDataType().size, column.getName());
      rpcColumns.add(rpcCol);
    }
    return rpcColumns;
  }

  /**
   * Converts an rpc schema to a schema for confluo
   *
   * @param rpcSchema The rpc schema to convert
   * @return The schema for confluo
   */
  static Schema convertToSchema(List<rpc_column> rpcSchema) {
    SchemaBuilder builder = new SchemaBuilder();
    for (rpc_column column : rpcSchema) {
      rpc_data_type type = rpc_data_type.findByValue(column.getTypeId());
      builder.addColumn(new DataType(type, column.getTypeSize()), column.getName());
    }
    return new Schema(builder.build());
  }
}

